File: src/j/implementation/microSet/InnerSet_4.java
package j.implementation.microSet;

import java.util.Collection;
import java.util.Iterator;

public class InnerSet_4 <T> extends AbstractInnerSet<T>
{
        protected T element_1;
        protected T element_2;
        protected T element_3;
        protected T element_4;
        
    public InnerSet_4(T p_element_1, T p_element_2, T p_element_3, T p_element_4)
    {
                element_1 = p_element_1;
                element_2 = p_element_2;
                element_3 = p_element_3;
                element_4 = p_element_4;
            }

    public InnerSet_4(InnerSet_3<T> i, T el)
    {
            element_1 = i.element_1;            element_2 = i.element_2;            element_3 = i.element_3;                    element_4 = el;
    }
    
    @Override
    public InnerSet<T> addElement(T p_element)
    {
        if(element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element) || element_4.equals(p_element))
            return this;
        else
            return new InnerSet_5<>(this, p_element);    }

    @Override
    public InnerSet<T> removeElement(Object p_element)
    {
                if(element_1.equals(p_element))
            return  new InnerSet_3<>(element_2, element_3, element_4);
         else                 if(element_2.equals(p_element))
            return  new InnerSet_3<>(element_1, element_3, element_4);
         else                 if(element_3.equals(p_element))
            return  new InnerSet_3<>(element_1, element_2, element_4);
         else                 if(element_4.equals(p_element))
            return  new InnerSet_3<>(element_1, element_2, element_3);
                        else
            return this;
    }

    @Override
    public boolean containsElement(Object p_element)
    {
        return element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element) || element_4.equals(p_element);
    }


    public T getElement(int index)
    {switch(index) {
                case 1:
            return element_1;
                case 2:
            return element_2;
                case 3:
            return element_3;
                case 4:
            return element_4;
                default:
            return null;
            }
    }


    public InnerSet<T> addAllElements(InnerSet<T> innerSet) {
                    innerSet = innerSet.addElement(element_1);
                    innerSet = innerSet.addElement(element_2);
                    innerSet = innerSet.addElement(element_3);
                    innerSet = innerSet.addElement(element_4);
                return innerSet;
    }


    @Override
    public InnerSet<T> removeAllElements(InnerSet<T> innerSet) {
                   innerSet = innerSet.removeElement(element_1);
                      innerSet = innerSet.removeElement(element_2);
                      innerSet = innerSet.removeElement(element_3);
                      innerSet = innerSet.removeElement(element_4);
                  return innerSet;
    }

    @Override
    public InnerSet<T> retainAllElements(InnerSet<T> innerSet) {
        InnerSet<T> i = this;
                if (!innerSet.containsElement(element_1)) i = i.removeElement(element_1);
                if (!innerSet.containsElement(element_2)) i = i.removeElement(element_2);
                if (!innerSet.containsElement(element_3)) i = i.removeElement(element_3);
                if (!innerSet.containsElement(element_4)) i = i.removeElement(element_4);
                return i;
    }

    @Override
    public boolean containsAllElements(InnerSet<T> innerSet) {
        return  innerSet.containsElement(element_1) && innerSet.containsElement(element_2) && innerSet.containsElement(element_3) && innerSet.containsElement(element_4);
    }

    @Override
    public String toString()
    {
        return "{" + element_1 + ", " + element_2 + ", " + element_3 + ", " + element_4 + " }";
    }
    
    @Override
    public int getSize()
    {
        return 4;
    }

    @Override
    public Iterator<T> getIterator() {
        return new InnerSetIterator<>(this);
    }

}

File: src/cpp/implementation/generate/inner_set_4.cpp
#include "inner_set_4.hpp"

template<typename T>
inner_set_4<T>::inner_set_4()
{

}

template<typename T>
inner_set_4<T>::inner_set_4(T p_element_1, T p_element_2, T p_element_3, T p_element_4)
{
            a_values.element_1 = p_element_1;
            a_values.element_2 = p_element_2;
            a_values.element_3 = p_element_3;
            a_values.element_4 = p_element_4;
    }

template<typename T>
inner_set_4<T>::inner_set_4(inner_set_3<T> p_inner_set, T p_element)
{
            a_values.element_1 = p_inner_set.a_values.element_1;            a_values.element_2 = p_inner_set.a_values.element_2;            a_values.element_3 = p_inner_set.a_values.element_3;                    a_values.element_4 = p_element;
}

template<typename T>
inner_set_4<T>::~inner_set_4()
{

}

                                                template<typename T>
inner_set<T>* inner_set_4<T>::add_element(T p_element)
{
    if(!contains_element(p_element))
        return new inner_set_5<T>(this, p_element);
    else
        return this;
}

template<typename T>
inner_set<T>* inner_set_4<T>::add_elements(T* p_elements)
{
    return nullptr; //TODO
}

template<typename T>
bool inner_set_4<T>::contains_element(T p_element)
{
    return  a_values.element_1 == p_element ||  a_values.element_2 == p_element ||  a_values.element_3 == p_element ||  a_values.element_4 == p_element;
}

template<typename T>
bool inner_set_4<T>::contains_all_elements(T* p_elements)
{
    return false; //TODO
}

template<typename T>
T inner_set_4<T>::get_element(int p_index)
{
    return reinterpret_cast<T*>(&a_values)[p_index];
}

template<typename T>
inner_set<T>* inner_set_4<T>::remove_element(T p_element)
{
            if(a_values.element_1 == p_element)
            return new inner_set_3<T>( a_values.element_2,  a_values.element_3,  a_values.element_4);
         else             if(a_values.element_2 == p_element)
            return new inner_set_3<T>( a_values.element_1,  a_values.element_3,  a_values.element_4);
         else             if(a_values.element_3 == p_element)
            return new inner_set_3<T>( a_values.element_1,  a_values.element_2,  a_values.element_4);
         else             if(a_values.element_4 == p_element)
            return new inner_set_3<T>( a_values.element_1,  a_values.element_2,  a_values.element_3);
                else
        return this;
}

template<typename T>
inner_set<T>* inner_set_4<T>::retain_all_element(T* p_elements)
{
    return this; //TODO
}

template<typename T>
int inner_set_4<T>::get_size()
{
    return 4;
}